### PR TITLE
refactor: refactor loss function

### PR DESCRIPTION
**Pull request by yuki-97**
1. Move parallel stuffs out of loss function.
2. Add `LossInputType` (logit, logprob, distillation) and `prepare_loss_input` to convert logits to the destination loss input and measure the parallel stuffs.
3. Update the loss file structure.
```
├── loss
│   ├── __init__.py
│   ├── interfaces.py
│   ├── loss_functions.py
│   ├── utils.py
│   └── wrapper.py
```

**Test Result**
https://wandb.ai/nvidia/refactor-loss-yukih?nw=0k8r2x613fml

| GRPO|
|-|
| image |

| SFT | Distillation |
|-|-|
| image | image |

| DPO | RM |
|-|-|
| image | image |

Nightly test all passed except the tests that already failed at main. https://github.com/NVIDIA-NeMo/RL/issues/2041

## Summary by CodeRabbit

* **Documentation**
  * Updated module import paths in guides and documentation to reflect reorganized loss function architecture.

* **Refactor**
  * Restructured loss function modules into a new hierarchical package with improved interfaces and abstractions. Introduced `LossInputType` enumeration for standardized loss function input specifications. Updated loss function signatures to use pre-computed log probabilities instead of raw logits. Reorganized loss utilities including input preparation and sequence packing wrappers.

### CODE REVIEW

File: nemo_rl/distributed/model_utils.py [hunk 1]
```diff
@@ -825,6 +825,50 @@ def get_logprobs_from_vocab_parallel_logits(
     )
 
 
+def get_logprobs_from_logits(
```
Comment (terrykong) [inline nemo_rl/distributed/model_utils.py]: wdyt about maybe coming up with a naming convention for these methods? at the moment it's kind of confusing to see:
* get_logprobs_from_vocab_parallel_logits
* get_logprobs_from_logits
* dtensor_from_parallel_logits_to_logprobs

at the very least it seems like some of these we should maybe make protected since we don't expect users to call them outside of this module

File: nemo_rl/distributed/model_utils.py [hunk 2]
```diff
@@ -982,6 +1026,173 @@ def gather_logits_at_global_indices(
     return gathered_logits
 
 
+def get_distilllation_topk_logprobs_from_logits(
```
Comment (terrykong) [inline nemo_rl/distributed/model_utils.py]:
```suggestion
def get_distillation_topk_logprobs_from_logits(
```

File: nemo_rl/models/automodel/train.py [hunk 3]
```diff
@@ -513,22 +523,65 @@ def __call__(
                 logits, self.device_mesh, self.cp_mesh, sequence_dim
             )
 
+        # Prepare data for loss function
+        def prepare_for_loss_fn(
+            logits: torch.Tensor, mb: BatchedDataDict[Any]
+        ) -> tuple[Any]:
+            if isinstance(self.loss_fn, (ClippedPGLossFn, NLLLoss, DPOLossFn)):
```
Comment (terrykong) [inline nemo_rl/models/automodel/train.py]: so i have a little mixed feelings about this one. since there's now a coupling between the loss functions here

is it possible to refactor so that each loss function has to define it's own "prepare"?

also, i can see this being a little error prone b/c these are positional and also located somewhere else apart from the loss function, so if the args are swapped (say teacher & student logits), it'd be hard to find such a bug. maybe kwargs or something at the very least

File: nemo_rl/algorithms/loss_functions.py [hunk 4]
```diff
@@ -925,22 +795,27 @@ class SequencePackingLossWrapper:
     def __init__(
         self,
         loss_fn: LossFunction,
+        prepare_fn: Callable[Any, Any],
         cu_seqlens_q: Tensor,
         cu_seqlens_q_padded: Optional[Tensor] = None,
+        vocab_parallel_rank: Optional[int] = None,
+        vocab_parallel_group: Optional[torch.distributed.ProcessGroup] = None,
```
Comment (terrykong) [inline nemo_rl/algorithms/loss_functions.py]: leaving as a todo since these aren't used b/c the mcore path hasn't incorporated these changes

File: nemo_rl/algorithms/loss_functions.py [hunk 5]
```diff
@@ -984,14 +859,14 @@ def __call__(
                 1, logit_start, logit_length
             )
 
+            # prepare data for loss function
+            loss_fn_args = self.prepare_fn(next_token_logits_slice, unpadded_seq_data)
+
             loss, metrics = self.loss_fn(
-                next_token_logits_slice,
+                *loss_fn_args,
```
Comment (terrykong) [inline nemo_rl/algorithms/loss_functions.py]: just another note on the positional args

**Review by terrykong — COMMENTED** [review]
@hemildesai can you review the automodel changes
@yaoyu-33 @cuichenx to comment on the change from the megatron side (although this PR hasn't implemented that part yet)
@zpqiu to comment on the distillation changes

File: nemo_rl/distributed/model_utils.py (hunk 1 above)
Comment (yuki-97) [reply nemo_rl/distributed/model_utils.py]: yea I also feel it's confused. but since this PR is a bit large and currently the usage is also confused (e.g. both `get_logprobs_from_vocab_parallel_logits` and `get_next_token_logprobs_from_logits` will be use outside, but `dtensor_from_parallel_logits_to_logprobs` only used at `model_utils.py`), I tend to clean it in a subsequent PR.
filed an issue here for tracking: https://github.com/NVIDIA-NeMo/RL/issues/2033, is it fine to you?

File: nemo_rl/algorithms/loss_functions.py (hunk 4 above)
Comment (yuki-97) [reply nemo_rl/algorithms/loss_functions.py]: https://github.com/NVIDIA-NeMo/RL/pull/1920/changes/2f1d5640bb72e70ce3948f21fa44c2eea05c6564

File: nemo_rl/models/automodel/train.py (hunk 3 above)
Comment (yuki-97) [reply nemo_rl/models/automodel/train.py]: as discussed offline, use `LossInputType`. https://github.com/NVIDIA-NeMo/RL/pull/1920/changes/0838ab461efb5366f01ff42e934afbda8620d16d

File: nemo_rl/algorithms/loss_functions.py (hunk 5 above)
Comment (yuki-97) [reply nemo_rl/algorithms/loss_functions.py]: https://github.com/NVIDIA-NeMo/RL/pull/1920/changes/1920542426979fd710284bc5cfe4f9db9c607062

File: nemo_rl/distributed/model_utils.py (hunk 2 above)
Comment (yuki-97) [reply nemo_rl/distributed/model_utils.py]: good catch! fixed.

**Comment by yuki-97** [comment]
hi @terrykong @hemildesai @yaoyu-33 @cuichenx @zpqiu, this PR should be ready for review (although there may still have some failing tests need to be fixed), can you help to take a review?

File: nemo_rl/algorithms/loss/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+class SequencePackingLossWrapper:
+    def __init__(
+        self,
+        loss_fn: LossFunction,
+        prepare_fn: Callable[Any, Any],
+        cu_seqlens_q: Tensor,
+        cu_seqlens_q_padded: Optional[Tensor] = None,
+        vocab_parallel_rank: Optional[int] = None,
+        vocab_parallel_group: Optional[torch.distributed.ProcessGroup] = None,
+        context_parallel_group: Optional[torch.distributed.ProcessGroup] = None,
```
Comment (terrykong) [inline nemo_rl/algorithms/loss/wrapper.py:36]: should this have a docstring to document that explains these are only used in the mcore backend and it's not a bug that it's missing in the automodel side

File: nemo_rl/algorithms/loss/interfaces.py
Comment (terrykong) [inline nemo_rl/algorithms/loss/interfaces.py]: the arguments/docstring should be updated. out of curiosity, did pyrefly catch catch this at lint time?

File: nemo_rl/algorithms/loss/wrapper.py (hunk 6 above)
Comment (yuki-97) [reply nemo_rl/algorithms/loss/wrapper.py:36]: https://github.com/NVIDIA-NeMo/RL/pull/1920/changes/443d7ad0205b656f083736e24ba7bb2a75def50a

File: nemo_rl/algorithms/loss/interfaces.py
Comment (yuki-97) [reply nemo_rl/algorithms/loss/interfaces.py]: https://github.com/NVIDIA-NeMo/RL/pull/1920/changes/443d7ad0205b656f083736e24ba7bb2a75def50a

yea, pyrefly didn't catch this :(

**Comment by yuki-97** [comment]
> Since this change touches every algorithm, can you run the nightlies?

yea, I've run the nightly tests, all tests passed except the tests that already failed at main, so this PR should be fine.

I filed an issue with error logs: https://github.com/NVIDIA-NeMo/RL/issues/2041. maybe assign someone to fix? @terrykong
